### PR TITLE
Update walk through and user guides

### LINKED ISSUE #1273: Walk-through documentation updates
#1271 is already possible but not obvious from the docs, https://hera.readthedocs.io/en/stable/examples/workflows/misc/global_config/ is the only example.

Currently, `set_class_defaults` itself is also not searchable in the docs:

image

And does not appear in the [Pre-build hooks section](https://hera.readthedocs.io/en/stable/walk-through/advanced-hera-features/#pre-build-hooks) - consider renaming the section to drop "advanced" and make a new walk through section for hera-specific features (i.e. things that are not a 1-1 mapping from Argo Workflows).

### PR DESCRIPTION
**Pull request by elliotgunton**
**Pull Request Checklist**
- [x] Fixes #1273 
- [ ] ~Tests added~ Docs only
- [x] Documentation/examples added
- [x] [Good commit messages](https://cbea.ms/git-commit/) and/or PR title

**Description of PR**
Main changes:
* Surface the "hera developer features" walk through section in the walk through
  before the "authentication" section so users start seeing Hera's unique features
  earlier
  * Added the "Set class defaults" section here
* Separate out the experimental features into its own section as it's big enough
* Rename the "advanced features" section to "further reading" to be less scary
* Remove the under-utilised Pydantic integration section, move relevant
  sections of text to the user guides - new "Integrated Pydantic Support" section
  in script-basics.md for the RunnerScriptConstructor
* Add more motivation to the Runner IO to make it more obvious why it's needed

### CODE REVIEW

**Comment by elliotgunton** [comment]
Draft until #1287 is merged (please review that one first, this PR consists of unrelated docs changes but builds on top of that branch).

File: docs/user-guides/script-runner-io.md [hunk 1]
```diff
@@ -1,8 +1,58 @@
...
+
+@script(constructor="runner")
+def my_function(my_input: MyInput) -> MyOutput:
+    ...
+    return MyOutput(
+        param_int_output=42,
+        another_param_int_output=-1,
+        a_str_param_output="Hello, world!",
+    )
+
```
Comment (alicederyn) [inline docs/user-guides/script-runner-io.md:51]:
```suggestion
```

File: docs/user-guides/script-runner-io.md [hunk 2]
```diff
@@ -1,8 +1,58 @@
...
+    )
+    artifact_int: Annotated[int, Artifact(name="artifact-input", loader=ArtifactLoader.json)]
+
+class MyOutput(Output):
+    param_int_output: Annotated[int, Parameter(name="param-int-output")],
+    another_param_int_output: Annotated[int, Parameter(name="another-param-int-output")],
+    a_str_param_output: Annotated[str, Parameter(name="a-str-param-output")],
+
+@script(constructor="runner")
+def my_function(my_input: MyInput) -> MyOutput:
```
Comment (alicederyn) [inline docs/user-guides/script-runner-io.md:44]: Should we maybe use an input parameter in this example, to demonstrate how that changes too?

File: mkdocs.yml [hunk 3]
```diff
@@ -18,12 +18,12 @@ nav:
...
     - walk-through/suspending.md
-    - Advanced Features:
-      - walk-through/authentication.md
-      - walk-through/advanced-workflow-features.md
-      - walk-through/advanced-template-features.md
-      - walk-through/advanced-hera-features.md
-      - walk-through/pydantic-support.md
+    - walk-through/hera-developer-features.md
+    - walk-through/authentication.md
+    - walk-through/experimental-hera-features.md
```
Comment (alicederyn) [inline mkdocs.yml:23]: Should this maybe be under "Further Reading"?

File: docs/user-guides/script-runner-io.md (hunk 2 above)
Comment (elliotgunton) [reply docs/user-guides/script-runner-io.md:44]: Do you mean without `Annotated`?

Comment (alicederyn) [reply docs/user-guides/script-runner-io.md:44]: I mean that a reference to `param_int` in the body of the original function would become `my_input.param_int` in the Pydantic-based one.

Comment (elliotgunton) [reply docs/user-guides/script-runner-io.md:45]: Done!

**Review by alicederyn — APPROVED** [review]
🚀

**Pull request merged**